src/magma/vk/textures/TextureManager.cpp

```diff
@@ -0,0 +1,213 @@
+    VkDeviceMemory textureMemory = _device.createDeviceMemory(memRequirements, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+
+    vkBindImageMemory(_device.handler(), textureImage, textureMemory, 0);
+
+    VkImageViewCreateInfo viewInfo{};
+    viewInfo.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+    viewInfo.image = textureImage;
+    viewInfo.viewType = VK_IMAGE_VIEW_TYPE_2D;
+    viewInfo.format = format;
+    viewInfo.subresourceRange.aspectMask = aspectMask;
+    viewInfo.subresourceRange.baseMipLevel = 0;
+    viewInfo.subresourceRange.levelCount = 1;
+    viewInfo.subresourceRange.baseArrayLayer = 0;
+    viewInfo.subresourceRange.layerCount = 1;
+
+    VkImageView textureView;
+    result = vkCreateImageView(_device.handler(), &viewInfo, nullptr, &textureView);
+    VK_CHECK_ERR(result, "TextureManager::failed to create image view!");
+
+    TextureInfo* textureInfo = new TextureInfo;
+    textureInfo->imageInfo = imageInfo;
+    textureInfo->viewInfo = viewInfo;
+    textureInfo->curLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+    textureInfo->name = name;
+
+    _textures.emplace(name, Texture(textureImage, textureMemory, ImageView(textureView), textureInfo));
+    return _textures.at(name);
+}
+
+static bool hasStencilComponent(VkFormat format) {
+    return format == VK_FORMAT_D32_SFLOAT_S8_UINT || format == VK_FORMAT_D24_UNORM_S8_UINT;
+}
+
+void TextureManager::setLayout(Texture &texture, VkImageLayout newLayout) {
+    VkImageCreateInfo imageInfo = texture.getInfo()->imageInfo;
+    _commandBuffers.resetCmdBuf(0);
+    VkCommandBuffer cmdBuf = _commandBuffers.beginCmdBuf(0);
+    {
+        VkImageLayout oldLayout = texture.getInfo()->curLayout;
+        VkImageMemoryBarrier barrier{};
+        barrier.sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+        barrier.oldLayout = oldLayout;
+        barrier.newLayout = newLayout;
+        barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+        barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+        barrier.image = texture.getImage();
+        barrier.subresourceRange.baseMipLevel = 0;
+        barrier.subresourceRange.levelCount = imageInfo.mipLevels;
+        barrier.subresourceRange.baseArrayLayer = 0;
+        barrier.subresourceRange.layerCount = imageInfo.arrayLayers;
+
+        VkPipelineStageFlags srcStage = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+        VkPipelineStageFlags dstStage = VK_PIPELINE_STAGE_ALL_COMMANDS_BIT;
+
+        if (newLayout == VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL)
+        {
+            barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
+
+            if (hasStencilComponent(imageInfo.format))
+                barrier.subresourceRange.aspectMask |= VK_IMAGE_ASPECT_STENCIL_BIT;
+        }
+        else
+        {
+            barrier.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+        }
+
+        switch (oldLayout) {
+        case VK_IMAGE_LAYOUT_UNDEFINED:
```

Comment:
I think all logical block inside the switch should have 1 extra tab

Reply:
Done